frame/support/src/storage/mod.rs

```diff
@@ -563,6 +572,138 @@ impl<T> Iterator for PrefixIterator<T> {
+	/// The prefix iterated on
+	prefix: Vec<u8>,
+	/// Storage key of the child trie
+	storage_key: Vec<u8>,
+	/// The last key iterated on
+	previous_key: Vec<u8>,
+	/// If true then values are removed while iterating
+	drain: bool,
+	/// Whether or not the first key i.e. the prefix is fetched
+	prefix_fetched: bool,
+	/// Function that takes `(raw_key_without_prefix, raw_value)` and decode `T`.
+	/// `raw_key_without_prefix` is the raw storage key without the prefix iterated on.
+	closure: fn(&[u8], &[u8]) -> Result<T, codec::Error>,
+}
+
+impl<T> ChildTriePrefixIterator<T> {
+	/// Mutate this iterator into a draining iterator; items iterated are removed from storage.
+	pub fn drain(mut self) -> Self {
+		self.drain = true;
+		self
+	}
+
+	/// Instruct the iterator to not skip the first key, i.e. the prefix.
+	pub fn fetch_prefix(mut self) -> Self {
+		self.prefix_fetched = false;
+		self
+	}
+}
+
+impl<T: Decode + Sized> ChildTriePrefixIterator<(Vec<u8>, T)> {
+	/// Construct iterator to iterate over child trie items in `child_info` with the prefix `prefix`.
+	/// 
+	/// By default the value under `prefix` is not fetched. Call the `fetch_prefix` method to enable
+	/// fetching the prefix.
+	pub fn with_prefix(child_info: &ChildInfo, prefix: &[u8]) -> Self {
+		let prefix = prefix.to_vec();
+		let previous_key = prefix.clone();
+		let closure = |raw_key_without_prefix: &[u8], raw_value: &[u8]| {
+			let value = T::decode(&mut &raw_value[..])?;
+			Ok((raw_key_without_prefix.to_vec(), value))
+		};
+
+		Self {
+			prefix,
+			storage_key: child_info.storage_key().to_vec(),
+			previous_key,
+			drain: false,
+			prefix_fetched: true,
+			closure,
+		}
+	}
+}
+
+impl<K: Decode + Sized, T: Decode + Sized> ChildTriePrefixIterator<(K, T)> {
+	/// Construct iterator to iterate over child trie items in `child_info` with the prefix `prefix`.
+	/// 
+	/// By default the value under `prefix` is not fetched. Call the `fetch_prefix` method to enable
+	/// fetching the prefix.
+	pub fn with_prefix_over_key<H: ReversibleStorageHasher>(child_info: &ChildInfo, prefix: &[u8]) -> Self {
+		let prefix = prefix.to_vec();
+		let previous_key = prefix.clone();
+		let closure = |raw_key_without_prefix: &[u8], raw_value: &[u8]| {
+			let mut key_material = H::reverse(raw_key_without_prefix);
+			let key = K::decode(&mut key_material)?;
+			let value = T::decode(&mut &raw_value[..])?;
+			Ok((key, value))
+		};
+
+		Self {
+			prefix,
+			storage_key: child_info.storage_key().to_vec(),
+			previous_key,
+			drain: false,
+			prefix_fetched: true,
+			closure,
+		 }
+	}
+}
+
+impl<T> Iterator for ChildTriePrefixIterator<T> {
+	type Item = T;
+
+	fn next(&mut self) -> Option<Self::Item> {
+		loop {
+			let maybe_next = if self.prefix_fetched {
+				sp_io::default_child_storage::next_key(&self.storage_key, &self.previous_key)
+					.filter(|n| n.starts_with(&self.prefix))
+			} else {
+				self.prefix_fetched = true;
+				Some(self.previous_key.clone())
```

Comment:
it is slightly unexpect to fetch the previous key when the field prefix_fetched is false.

Maybe we should rename the field to `fetch_previous_key`
if `fetch_previous_key` is true then we need to fetch previous_key and we set to false
if it is false then we just take the next_key